### PR TITLE
feat!: support full_abi in decode_returndata

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

Include full ABI as optional argument when decoding return data.
This is needed in some ecosystems where struct types are defined elsewhere in the ABI.

### How I did it

Refactoring mostly!

### How to verify it

can get https://github.com/ApeWorX/ape-starknet/pull/69 to work

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape/api/networks.py [hunk 1]
```diff
@@ -342,13 +342,15 @@ def create_transaction(self, **kwargs) -> "TransactionAPI":
         """
 
     @abstractmethod
-    def decode_returndata(self, abi: MethodABI, raw_data: bytes) -> Any:
+    def decode_returndata(self, abi: MethodABI, raw_data: bytes, full_abi: List[ABI]) -> Any:
```
Comment (ca11ab1e) [inline src/ape/api/networks.py:345]: Should be optional.

File: src/ape/api/networks.py [hunk 2]
```diff
@@ -342,13 +342,15 @@ def create_transaction(self, **kwargs) -> "TransactionAPI":
...
         """
         Get the result of a contract call.
 
         Arg:
             abi (MethodABI): The method called.
             raw_data (bytes): Raw returned data.
+            address (Optional[AddressType]): The address of the contract, if needed.
```
Comment (ca11ab1e) [inline src/ape/api/networks.py:352]: To update.

**Comment by ca11ab1e** [comment]
I'll be able to test tomorrow (UTC+2).

File: src/ape/api/networks.py (hunk 1 above)
Comment (ca11ab1e) [inline src/ape/api/networks.py:345]:
```suggestion
    def decode_returndata(self, abi: MethodABI, raw_data: bytes, full_abi: Optional[List[ABI]] = None) -> Any:
```

File: src/ape/api/networks.py [hunk 3]
```diff
@@ -342,13 +342,15 @@ def create_transaction(self, **kwargs) -> "TransactionAPI":
...
+              Useful for looking up additional ABIs. Defaults to None.
```
Comment (ca11ab1e) [inline src/ape/api/networks.py:353]:
```suggestion
            full_abi: Optional[List[ABI]] = None: The complete contract's ABI, if needed.
              Useful for decoding transactions return data. Defaults to None.
```

File: src/ape/utils/trace.py [hunk 4]
```diff
@@ -151,7 +151,9 @@ def _dim_default_gas(call_sig: str) -> str:
                 # The revert-message appears at the top of the trace output.
                 try:
                     return_value = (
-                        self.decode_returndata(method, call.returndata) if not call.failed else None
+                        self.decode_returndata(method, call.returndata, contract_type.abi)
```
Comment (ca11ab1e) [inline src/ape/utils/trace.py:154]:
```suggestion
                        self.decode_returndata(method, call.returndata, full_abi=contract_type.abi)
```

File: src/ape/utils/trace.py [hunk 5]
```diff
@@ -235,8 +237,11 @@ def decode_calldata(self, method: MethodABI, raw_data: bytes) -> Dict:
 
         return arguments
 
-    def decode_returndata(self, method: MethodABI, raw_data: bytes) -> Any:
-        values = [self.decode_value(v) for v in self._ecosystem.decode_returndata(method, raw_data)]
+    def decode_returndata(self, method: MethodABI, raw_data: bytes, full_abi: List[ABI]) -> Any:
```
Comment (ca11ab1e) [inline src/ape/utils/trace.py:240]:
```suggestion
    def decode_returndata(self, method: MethodABI, raw_data: bytes, full_abi: Optional[List[ABI]] = None) -> Any:
```

File: src/ape/utils/trace.py [hunk 6]
```diff
@@ -235,8 +237,11 @@ def decode_calldata(self, method: MethodABI, raw_data: bytes) -> Dict:
...
+        values = [
+            self.decode_value(v)
+            for v in self._ecosystem.decode_returndata(method, raw_data, full_abi)
```
Comment (ca11ab1e) [inline src/ape/utils/trace.py:243]:
```suggestion
            for v in self._ecosystem.decode_returndata(method, raw_data, full_abi=full_abi)
```

File: src/ape_ethereum/ecosystem.py [hunk 7]
```diff
@@ -267,7 +267,9 @@ def encode_calldata(self, abi: Union[ConstructorABI, MethodABI], *args) -> bytes
         else:
             return HexBytes(b"")
 
-    def decode_returndata(self, abi: MethodABI, raw_data: bytes) -> Tuple[Any, ...]:
+    def decode_returndata(
+        self, abi: MethodABI, raw_data: bytes, full_abi: List[ABI]
```
Comment (ca11ab1e) [inline src/ape_ethereum/ecosystem.py:271]:
```suggestion
        self, abi: MethodABI, raw_data: bytes, full_abi: Optional[List[ABI]] = None
```

**Review by ca11ab1e — APPROVED** [review]

**Comment by unparalleled-js** [comment]
Talked with @fubuloubu , new plan!
We are going to try to upgrade ethpm-types to be able to hold back refs to their contract type and have contract type parsing automatically set the references.

**Pull request closed without merging**